### DISCUSSION TITLE
How to use RP2350 hardware TRNG?

### DISCUSSION THREAD

**Discussion by metehoca** [earlephilhower/arduino-pico#2643] [Q&A] [answered]
The RP2350 is stated to include a hardware-based True Random Number Generator (TRNG), but I couldn't find an example specific to its use. Most examples available are based on Ring Oscillators (ROSC), which often produce lower-quality randomness compared to a true TRNG and are less reliable than even Arduino's random() function. Is there a way to utilize the RP2350's TRNG through the Arduino-Pico library?

**Comment by maxgerhardt** [comment]
The TRNG is supported through Pico-SDK's [pico_rand](https://github.com/raspberrypi/pico-sdk/blob/95ea6acad131124694cda1c162c52cd30e0aece0/src/rp2_common/pico_rand/rand.c) module, which we do include in our precompiled libpico [build configuration](https://github.com/earlephilhower/arduino-pico/blob/e16c459598b86ce67d8b83ac5b1384dcb8c982ae/tools/libpico/CMakeLists.txt#L192). The module uses the TRNG output to feed a xoroshiro128 generator, along a lot of other sources of randomness, e.g., the ROSC, the system time, the random initial values of SRAM or the board's unique ID. 

There is also Arduino wrapper library for this module:

https://github.com/earlephilhower/arduino-pico/blob/e16c459598b86ce67d8b83ac5b1384dcb8c982ae/cores/rp2040/RP2040Support.h#L463-L465

That means that `uint32_t rand_value = rp2040.hwrand();` will work.

```cpp
#include <Arduino.h>

void setup() {
  Serial.begin(115200);
}

void loop() {
  uint32_t random_val = rp2040.hwrand32();
  Serial.println("Random value: 0x" + String(random_val, HEX));
  delay(1000);
}
```

We can call also the functions of the pico_rand module directly. You can also print out which exact sources were used.

```cpp
#include <Arduino.h>
#include <pico/rand.h>

void setup() {
  Serial.begin(115200);
}

String get_used_rng_sources() {
  String used_srcs = "";
#ifdef PICO_RAND_ENTROPY_SRC_TIME
  used_srcs += "time,";
#endif
#if PICO_RAND_ENTROPY_SRC_ROSC
  used_srcs += "rosc,";
#endif
#if PICO_RAND_ENTROPY_SRC_TRNG
  used_srcs += "trng,";
#endif
#if PICO_RAND_ENTROPY_SRC_BUS_PERF_COUNTER
  used_srcs += "perf_count,";
#endif
  return used_srcs;
}

void loop() {
  uint32_t random_val = get_rand_32();
  Serial.println("RNG sources: " + get_used_rng_sources() + ". Random value: 0x" + String(random_val, HEX));
  delay(1000);
}
```

Prints on a RP2350 based board:
```
RNG sources: time,trng. Random value: 0x8867453a
RNG sources: time,trng. Random value: 0x3c41e893
..
```

**Reply by maxgerhardt** [reply, marked as answer]
Note that if you want to get the TRNG output directly, you can also do that.

```cpp
#include <Arduino.h>
#include "hardware/structs/trng.h"

void setup() {
  Serial.begin(115200);
  // initially reset TRNG
  trng_hw->trng_sw_reset = 1;
  delayMicroseconds(100);
  trng_hw->trng_sw_reset &= ~1;
}

uint32_t trng_sample_words[count_of(trng_hw->ehr_data)];
static_assert(count_of(trng_hw->ehr_data) >= 2 && count_of(trng_hw->ehr_data) < 255, "");
uint8_t trng_sample_word_count;

static uint64_t trng_get_64bits(void) {
  spin_lock_t *lock = spin_lock_instance(PICO_SPINLOCK_ID_RAND);
  uint32_t save = spin_lock_blocking(lock);
  if (trng_sample_word_count < 2) {
    // Sets the number of rng_clk cycles between two consecutive ring oscillator samples
    // the higher this is, the more time it will take to generate a random number.
    trng_hw->sample_cnt1 = 0x0000ffff; 
    trng_hw->trng_config = 3; // number of inverters to use
    trng_hw->trng_debug_control = 0;
    trng_hw->rnd_source_enable |= 1;
    // Clear all interrupts (including EHR_VLD)
    trng_hw->rng_icr = -1u;
    while (!trng_hw->trng_valid); // wait for all valid data
    for (uint i = 0; i < count_of(trng_sample_words); i++) {
      trng_sample_words[i] = trng_hw->ehr_data[i];
    }
    trng_sample_word_count = count_of(trng_sample_words);
    // trng_hw->rnd_source_enable = 0; // disable clock again if you want to save power
  }
  trng_sample_word_count -= 2;
  uint64_t rc = trng_sample_words[trng_sample_word_count] |
                (((uint64_t)trng_sample_words[trng_sample_word_count + 1]) << 32);
  spin_unlock(lock, save);
  return rc;
}

void loop() {
  uint64_t random_val = trng_get_64bits();
  char buf[(64 / 8 * 2) + 1] = {};
  Serial.println("Random value: 0x" + String(ulltoa(random_val, buf, sizeof(buf), 16)));
  delay(1000);
}
```
just be sure to read all the caveats for this such as long / unpredictable generation times from the RP2350 datasheet. And of course, that code would be RP2350 only.

```
Random value: 0xf45803320b955f27
Random value: 0x576d4944d971afe4
Random value: 0x5a8c3907cef6f120
```

So really, if you have no special requirements except for "it must have used the TRNG to at least seed a good pseudo-random number generator", `rp2040.hwrand32()` aka `get_rand_32()` from `pico_rand` will work just fine on RP2350, and still do its best on RP2040 which has no TRNG.
